src/python/pants/backend/jvm/rules/coursier.py

```diff
@@ -0,0 +1,342 @@
+    org_name_to_org_name_rev[f'{coord.org}:{coord.name}'] = coord
+
+  jars_per_target = []
+
+  override_classifiers = CoursierMixin.override_classifiers_for_conf(req.conf.conf)
+
+  for ht in req.source_targets:
+    t = ht.adaptor
+    jars_to_digest = []
+    if isinstance(t.v1_target, JarLibrary):
+      def get_transitive_resolved_jars(my_coord, resolved_jars):
+        transitive_jar_path_for_coord = []
+        coord_str = str(my_coord)
+        if coord_str in flattened_resolution and my_coord in resolved_jars:
+          transitive_jar_path_for_coord.append(resolved_jars[my_coord])
+
+          for c in flattened_resolution[coord_str]:
+            j = resolved_jars.get(CoursierMixin.to_m2_coord(c))
+            if j:
+              transitive_jar_path_for_coord.append(j)
+
+        return transitive_jar_path_for_coord
+
+      for jar in t.jar_dependencies:
+        # if there are override classifiers, then force use of those.
+        coord_candidates = []
+        if override_classifiers:
+          coord_candidates = [jar.coordinate.copy(classifier=c) for c in override_classifiers]
+        else:
+          coord_candidates = [jar.coordinate]
+
+        # If there are conflict resolution entries, then update versions to the resolved ones.
+        jar_spec = f'{jar.coordinate.org}:{jar.coordinate.name}'
+        if jar.coordinate.simple_coord in parsed_output['conflict_resolution']:
+          parsed_conflict = CoursierMixin.to_m2_coord(
+            parsed_output['conflict_resolution'][jar.coordinate.simple_coord])
+          coord_candidates = [c.copy(rev=parsed_conflict.rev) for c in coord_candidates]
+        elif jar_spec in org_name_to_org_name_rev:
+          parsed_conflict = org_name_to_org_name_rev[jar_spec]
+          coord_candidates = [c.copy(rev=parsed_conflict.rev) for c in coord_candidates]
+
+        for coord in coord_candidates:
+          transitive_resolved_jars = get_transitive_resolved_jars(coord, coord_to_resolved_jars)
+          if transitive_resolved_jars:
+            for jar in transitive_resolved_jars:
+              jars_to_digest.append(jar)
+
+      jars_per_target.append((t.v1_target, tuple(jars_to_digest)))
+
+
+  return CoursierResolveResult(
+    base_conf=req.conf,
+    jars_per_target=tuple(jars_per_target),
+  )
+
+
+@dataclass(frozen=True)
+class SnapshottedResolveResult:
+  resolution: CoursierResolveResult
+  merged_snapshot: Snapshot
+
+  def __getattr__(self, key, **kwargs):
+    """Use the prototype pattern with the original coursier resolution resuslt."""
+    if hasattr(self.resolution, key):
+      return getattr(self.resolution, key)
+    return super().__getattr__(key, **kwargs)
+
+
+@rule
+async def snapshotted_coursier_result(
+    res: CoursierResolveResult,
+    coursier: CoursierSubsystem,
+    build_root: BuildRoot,
+) -> SnapshottedResolveResult:
+  # TODO: figure out whether the arbitrary_root kwarg is the right way to snapshot things outside of
```

Comment:
There is no safe way to do this with the engine currently, because we can't watch arbitrary paths. Perhaps after #4999